upstairs/src/volume.rs

```diff
@@ -1238,14 +1234,130 @@ impl Volume {
+                    ))
+                }
+            } else {
+                // New read_only_parent is none, so the read_only_parents can't
+                // be considered for replacement.
+                Err(CrucibleError::ReplaceRequestInvalid(
+                    "read_only_parent does not exist in replacement"
+                        .to_string(),
+                ))
+            }
+        };
+
+        // Determine the sub_volume difference if there is any.
+        let sub_volume_result = Self::compare_vcr_region_for_replacement(
+            &o_sub_volumes[0],
+            &n_sub_volumes[0],
+        );
+        info!(
+            log,
+            "sub_volume_result is: {:?}, ROP:{:?}",
+            sub_volume_result,
+            rop_result
+        );
+
+        // So, now time for the truth.
+        // We use the same function to compare both the region in the
+        // sub_volumes and the region in the read_only_parent.  Only one
+        // can be different, so we walk the cases.
+        match sub_volume_result {
+            Ok(Some((old, new))) => {
+                // The sub_volume had a valid difference, if the
+                // read_only_parent also had a difference, then that is an
+                // error, otherwise, we return what the sub_volume found.
+                match rop_result {
+                    Ok(Some((_, _))) => {
+                        // Both have changes, this is no good
+                        crucible_bail!(
+                            ReplaceRequestInvalid,
+                            "Both sub_volume and read_only_parent have changes"
+                        )
+                    }
+                    _ => Ok(Some((old, new))),
+                }
+            }
+            Ok(None) => {
+                // The sub_volume VCR is the same as what we have, but with
+                // updated generation numbers, so this qualifies for the
+                // Ok(None) state.  Check what the ROP found.
+                match rop_result {
+                    Ok(Some((read_only_old, read_only_new))) => {
+                        // Just the ROP has changes, so let's return those.
+                        Ok(Some((read_only_old, read_only_new)))
+                    }
+                    _ => Ok(None),
+                }
+            }
+            Err(e) => {
+                // The sub_volume replacement is not acceptable.  Check if
+                // the read_only_parent has a valid change.  If not, then
+                // return the error from the sub_volume.
+                match rop_result {
+                    Ok(Some((ro, rn))) => {
+                        // Just the ROP has changes, return those.
+                        Ok(Some((ro, rn)))
+                    }
+                    Ok(None) => Ok(None),
+                    Err(_) => Err(e),
+                }
+            }
+        }
```

Comment:
I'd recommend matching against the tuple of `(sub_volume_result, rop_result)`, to make more clear what cases are being handled.  Here's my translation of the current code:
```suggestion
        // So, now time for the truth.
        // We use the same function to compare both the region in the
        // sub_volumes and the region in the read_only_parent.  Only one
        // can be different, so we walk the cases.
        match (sub_volume_result, rop_result) {
            (Ok(Some(_)), Ok(Some(_))) => {
                // Both have changes, this is no good
                crucible_bail!(
                    ReplaceRequestInvalid,
                    "Both sub_volume and read_only_parent have changes"
                )
            }
            (r @ Ok(Some(_)), _) => {
                // Return what the subvolume found
                r
            }
            // The sub_volume VCR is the same as what we have, but with
            // updated generation numbers, so this qualifies for the
            // Ok(None) state.  Check what the ROP found.
            (Ok(None), r @ Ok(Some(_))) => {
                // Just the ROP has changes, so let's return those.
                r
            }
            (Ok(None), _) => Ok(None),

            // The sub_volume replacement is not acceptable.  Check if
            // the read_only_parent has a valid change.  If not, then
            // return the error from the sub_volume.
            (Err(_), r @ Ok(_)) => r,
            (Err(e), _) => Err(e),
        }
```

Reply:
This got a little more complex with the refactor.  There are now 4 possible
return states.  It's possible though that the tuple match is still more clear than
what I ended up with in the refactor.  Take a look and tell me what you think here.

Reply:
Okay, I'll read the new code and start a new thread